accepted/future-releases/variance/feature-specification.md

```diff
@@ -0,0 +1,373 @@
+
+This feature introduces explicit variance modifiers for type parameters. It
+includes compile-time restrictions on type declarations and on the use of
+objects whose static type includes these modifiers, ensuring that the
+above-mentioned dynamic type errors cannot occur.
+
+In order to ease the transition where types with explicit variance are
+created and used, this proposal allows for certain subtype relationships
+where dynamic type checks are still needed when using legacy types (where
+type parameters are _implicitly_ covariant) to access an object, even in
+the case where the object has a type with explicit variance. For example,
+it is allowed to declare `class MyList<out E> implements List<E> {...}`,
+even though this means that `MyList` has members such as `add` that require
+dynamic checks and may incur a dynamic type error.
+
+
+## Syntax
+
+The grammar is adjusted as follows:
+
+```
+<typeParameter> ::= // Modified rule.
+    <metadata> <typeParameterVariance>? <typeIdentifier>
+    ('extends' <typeNotVoid>)?
+
+<typeParameterVariance> ::= // New rule.
+    'out' | 'inout' | 'in'
+```
+
+`out` and `inout` are added to the set of built-in identifiers (* and `in`
+is already a reserved word*).
+
+
+## Static Analysis
+
+This feature allows type parameters to be declared with a _variance
+modifier_ which is one of `out`, `inout`, or `in`. This implies that the
+use of such a type parameter is restricted, in return for improved static
+type safety. Moreover, the rules for other topics like subtyping and for
+determining the variance of a subterm in a type are adjusted.
+
+
+### Subtype Rules
+
+The interface compositionality rule in [subtyping.md] is updated
+as follows:
+
+[subtyping.md]: https://github.com/dart-lang/language/blob/master/resources/type-system/subtyping.md
+
+- **Interface Compositionality**: `T0` is an interface type `C0<S0, ..., Sk>`
+  and `T1` is `C0<U0, ..., Uk>`. For `i` in `0..k`, let `vi` be the declared
+  variance of the `i`th type parameter of `C0`. Then, for each `i` in `0..k`,
+  one of the following holds:
+  - `Si <: Ui` and `vi` is absent or `out`.
+  - `Ui <: Si` and `vi` is `in`.
+  - `Si <: Ui` and `Ui <: Si`, and `vi` is `inout`.
+
+
+### Variance Rules
+
+The rules for determining the variance of a position are updated as follows:
+
+We say that a type parameter of a generic class is _covariant_ if it has no
+variance modifier or it has the modifier `out`; we say that it is
+_contravariant_ if it has the modifier `in`; and we say that it is
+_invariant_ if it has the modifier `inout`.
+
+The covariant occurrences of a type (schema) `T` in another type (schema)
+`S` are:
+
+  - if `S` and `T` are the same type,
+    - `S` is a covariant occurrence of `T`.
+  - if `S` is `Future<U>`
+    - the covariant occurrences of `T` in `U`
+  - if `S` is `FutureOr<U>`
+    - the covariant occurrencs of `T` in `U`
+  - if `S` is an interface type `C<T0, ..., Tk>`
```

Comment:
I think this is ok, but it feels slightly off to me that we don't consider an invariant occurrence to be both a covariant occurrence and an invariant occurrence.  I think it works out with the way the definitions are used here?  But it's worth double checking

Reply:
I think it's working (it's the same approach as in the language specification, and I think that's in a good shape).